### PR TITLE
Fixed #29396 -- Fixed crash when using __year lookup with OuterRef.

### PR DESCRIPTION
**Pull request by Windsooon**
https://code.djangoproject.com/ticket/29396

### CODE REVIEW

**Comment by mitya57** [comment]
This patch seems to fix my issue, thanks! :+1:

**Comment by timgraham** [comment]
`tests/lookup` might be the place for the test. Try to reuse existing models if possible as adding models increases the time it takes to run all the tests.

**Comment by Windsooon** [comment]
@timgraham Thank you, please review the pr.

File: tests/lookup/models.py [hunk 1]
```diff
@@ -19,6 +20,7 @@ def __str__(self):
 class Author(models.Model):
     name = models.CharField(max_length=100)
     alias = models.CharField(max_length=50, null=True, blank=True)
+    date = models.DateField(default=datetime.date.today())
```
Comment (timgraham) [inline tests/lookup/models.py]: Remove `()` -- that's a fixed default which gives a check warning when running the tests.

**Comment by Windsooon** [comment]
I don't know how to do with this code

    `Author.objects.filter(date__year__gte=OuterRef("pub_date"))`

We only want to compare the year but we can't just get the year from `pub_date` with OuterRef,  right?

**Comment by charettes** [comment]
@Windsooon you could if the outer query annotated `pub_date_year=TruncYear('pub_date')` and you referred to it using `(date__year__gte=OuterRef('pub_date_year'))`.

**Comment by mitya57** [comment]
> The test seems to pass without the fix?

Yes, because `self.assertTrue(years.query)` does not actually compile the query to SQL. @Windsooon You need to use any of methods evaluating the queryset, or assert `str(years.query)`

> When I look at the sql, I found that the filter didn't work.

I do not understand what you are trying to achieve by putting the filter inside `TruncYear`. As far as I am aware, only the `Count` aggregate supports `filter` keyword.

I don’t think there is a way to make an alternative queryset to compare with. You can use `TruncYear`/`ExtractYear` to annotate the year, but it will still work only with the fix:

```python
Author.objects.annotate(date_year=ExtractYear("date")) \
              .filter(date_year__gte=OuterRef("pub_date"))
```
is identical to
```python
Author.objects.filter(date__year__gte=OuterRef("pub_date"))
```

**Comment by Windsooon** [comment]
Thank you @mitya57, so what we can do now is 
` self.assertEqual(str(year.query) == SQL)`?

**Comment by mitya57** [comment]
`self.assertEqual(str(year.query), sql)` sounds fine as long as the test is only run on one particular DB backend.

**Comment by timgraham** [comment]
I'd use an assertion for the results of the query (as the majority of Django's tests do) rather than the SQL.

**Comment by Windsooon** [comment]
I agreed with Tim. So I add another field for testing.

**Comment by Windsooon** [comment]
@timgraham anything I can do to speed up the merge? the checks didn't pass give 404 not found.

File: tests/lookup/tests.py [hunk 2]
```diff
@@ -909,3 +911,16 @@ def test_custom_lookup_none_rhs(self):
         field = query.model._meta.get_field('nulled_text_field')
         self.assertIsInstance(query.build_lookup(['isnull_none_rhs'], field, None), IsNullWithNoneAsRHS)
         self.assertTrue(Season.objects.filter(pk=season.pk, nulled_text_field__isnull_none_rhs=True))
+
+    def test_outerref_in_dates(self):
+        dates = Author.objects.filter(date__year__gte=OuterRef("year"))
```
Comment (timgraham) [inline tests/lookup/tests.py]: Use single quotes (unless a string contains a single quote) as described in [Python coding style](https://docs.djangoproject.com/en/dev/internals/contributing/writing-code/coding-style/#python-style).

File: tests/lookup/tests.py [hunk 3]
```diff
@@ -909,3 +911,16 @@ def test_custom_lookup_none_rhs(self):
...
+        dates_subq = Subquery(dates.values("id"))
+        years = Article.objects.filter(author__in=dates_subq)
+        self.assertQuerysetEqual(
+            years,
+            ['<Article: Article 1>', '<Article: Article 2>',
```
Comment (timgraham) [inline tests/lookup/tests.py]: Use the indentation style of the other tests. Actually, you can use `assertCountEqual()` instead. Maybe it makes sense to create another Article with a different year to test that the results are correct and not just returning everything.

File: django/db/models/lookups.py [hunk 4]
```diff
@@ -487,13 +487,21 @@ def year_lookup_bounds(self, connection, year):
...
         # We will need to skip the extract part and instead go
         # directly with the originating field, that is self.lhs.lhs.
         lhs_sql, params = self.process_lhs(compiler, connection, self.lhs.lhs)
         rhs_sql, rhs_params = self.process_rhs(compiler, connection)
         rhs_sql = self.get_rhs_op(connection, rhs_sql)
+        try:
+            # Check that rhs_params[0] exists (IndexError),
+            # it isn't None (TypeError), and is a number (ValueError)
+            int(rhs_params[0])
+        except (IndexError, TypeError, ValueError):
```
Comment (timgraham) [inline django/db/models/lookups.py]: It looks like only `IndexError` is exercised in tests.

**Comment by mitya57** [comment]
@Windsooon Can you please address the review comments?

**Comment by Windsooon** [comment]
@mitya57 I'm sorry I messed up with Git, I thought I updated the PR two months ago but it didn't :<. I push my PR, hope this would help.

**Comment by mitya57** [comment]
Now the diff contains only tests changes. Why did you drop the actual fix part?

**Comment by Windsooon** [comment]
@mitya57 Another mistakes I made :D

**Comment by mitya57** [comment]
Now I wonder why the tests are running against your previous commit. E.g. [this log](https://djangoci.com/job/pull-requests-bionic/database=mysql,label=bionic-pr,python=python3.5/3339/consoleText) has this line:

    Commit message: "Merge f155e28ec853b803ed8ddf752470baa9f09e5ebf into 573f44d62fe1e87e2c20a74eba5e20ca9ff0ed85"

The first commit ID is from your previous push, not from the current one.

Maybe one of the project members can restart the tests? GitHub's `refs/pull/10023/merge` is now up to date, so a new run should pick up the latest commit.

**Comment by mitya57** [comment]
@timgraham @charettes Ping, can you please retry tests and review this?

File: django/db/models/lookups.py (hunk 4 above)
Comment (timgraham) [reply django/db/models/lookups.py]: As far as I tested, this comment isn't addressed.

**Comment by Windsooon** [comment]
@timgraham anything I can do with the PR?

**Comment by charettes** [comment]
@Windsooon, Tim isn't available but I think you should start by addressing the comment he left back in November.

https://github.com/django/django/pull/10023/files#r252316281

https://github.com/django/django/pull/10023/files#r232433323

**Comment by Windsooon** [comment]
@charettes Thanks, I found I forget to submit my comments :<. Weird design by Github, right.

**Comment by charettes** [comment]
FWIW this issue doesn't have anything to do with subqueries/`OuterRef`. The `__year` lookups simply never added support for non-direct/expression values. The following test diff reproduces the issue without the use of subqueries

```diff
diff --git a/tests/db_functions/datetime/test_extract_trunc.py b/tests/db_functions/datetime/test_extract_trunc.py
index be1f36c89f..fe724ee41d 100644
--- a/tests/db_functions/datetime/test_extract_trunc.py
+++ b/tests/db_functions/datetime/test_extract_trunc.py
@@ -135,6 +135,10 @@ class DateFunctionTests(TestCase):
                 qs = DTModel.objects.filter(**{'start_datetime__%s__gte' % lookup: 2015})
                 self.assertEqual(qs.count(), 2)
                 self.assertEqual(str(qs.query).lower().count('extract'), 0)
+                qs = DTModel.objects.annotate(
+                    start_year=ExtractYear('start_datetime'),
+                ).filter(**{'end_datetime__%s__gte' % lookup: F('start_year')})
+                self.assertEqual(qs.count(), 2)

     def test_extract_year_lessthan_lookup(self):
         start_datetime = datetime(2015, 6, 15, 14, 10)
```

**Comment by felixxm** [comment]
Superseded by #11393.

**Pull request closed without merging**